File: src/test/java/dark/leech/text/action/export/ToCTest.java
package dark.leech.text.action.export;

import dark.leech.text.models.Chapter;
import dark.leech.text.models.Properties;
import org.junit.Test;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

import dark.leech.text.util.SettingUtils;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class ToCTest {

    @Test
    public void testNavPointClosingTags() throws Exception {
        Path tempDir = Files.createTempDirectory("leech-text-test-");
        
        // Create required data directories that caller is expected to provide
        new File(tempDir.toFile(), "data").mkdirs();
        new File(tempDir.toFile(), "data/Text").mkdirs();
        new File(tempDir.toFile(), "data/Images").mkdirs();
        
        // Initialize HTML_SYNTAX for tests because Settings dependencies may not load automatically
        SettingUtils.HTML_SYNTAX = "<html><head><title>Test</title></head><body></body></html>";
        
        Properties properties = new Properties();
        properties.setName("Test Book");
        properties.setAuthor("Test Author");
        properties.setSavePath(tempDir.toAbsolutePath().toString());
        properties.setAddGt(true); // Include Introduction to test its tags too
        
        List<Chapter> chapters = new ArrayList<>();
        chapters.add(new Chapter("url1", 1, "", "Chương 1: Bắt đầu"));
        chapters.add(new Chapter("url2", 2, "", "Chương 2: Tiếp tục"));
        properties.setChapList(chapters);
        
        ToC toc = new ToC(properties);
        toc.setAutoSplit(true);
        toc.setIncludeImg(false);
        
        // Generate ToC
        toc.mkToC();
        
        // Validate generated ncx file
        File ncxFile = new File(tempDir.toFile(), "data/toc.ncx");
        assertTrue("toc.ncx should be created", ncxFile.exists());
        
        String ncxContent = Files.readString(ncxFile.toPath());
        
        // We expect exact closing tags for navPoint 
        // We just verify that the number of opening <navPoint> matches </navPoint>
        
        int openCount = ncxContent.split("<navPoint").length - 1;
        int closeCount = ncxContent.split("</navPoint>").length - 1;
        
        assertEquals("All <navPoint> elements must be properly closed", openCount, closeCount);
    }
}

File: engine/src/jvmTest/kotlin/dev/haipham22/leechtext/action/export/ToCTest.kt
package dev.haipham22.leechtext.action.export

import dev.haipham22.leechtext.log.platformEngineLogger
import dev.haipham22.leechtext.models.Chapter
import dev.haipham22.leechtext.models.Properties
import dev.haipham22.leechtext.util.AppSettings
import dev.haipham22.leechtext.util.SettingsRepository
import java.io.File
import kotlin.io.path.createTempDirectory
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/** Port từ src/test/java/dark/leech/text/action/export/ToCTest.java. */
class ToCTest {
    private val settings: AppSettings = SettingsRepository.defaults()

    @Test
    fun navPointClosingTagsBalanceWithIntroSection() {
        val tempDir = createTempDirectory("leech-text-test-").toFile()

        // Caller phải tạo sẵn data dirs
        File(tempDir, "data").mkdirs()
        File(tempDir, "data/Text").mkdirs()
        File(tempDir, "data/Images").mkdirs()

        val properties =
            Properties().apply {
                name = "Test Book"
                author = "Test Author"
                savePath = tempDir.absolutePath
                addGt = true
                chapList =
                    listOf(
                        Chapter(url = "url1", chapName = "Chương 1: Bắt đầu").apply { id = "C1" },
                        Chapter(url = "url2", chapName = "Chương 2: Tiếp tục").apply { id = "C2" },
                    )
            }

        ToC(platformEngineLogger(), properties, autoSplit = true, includeImg = false, settings = settings).mkToC()

        val ncxFile = File(tempDir, "data/toc.ncx")
        assertTrue(ncxFile.exists(), "toc.ncx should be created")

        val ncxContent = ncxFile.readText()
        val openCount = ncxContent.split("<navPoint").size - 1
        val closeCount = ncxContent.split("</navPoint>").size - 1

        assertEquals(openCount, closeCount, "All <navPoint> elements must be properly closed")
    }

    @Test
    fun singlePartModeWritesManifestItemsForAllChapters() {
        val tempDir = createTempDirectory("leech-toc-single-").toFile()
        File(tempDir, "data").mkdirs()
        File(tempDir, "data/Text").mkdirs()

        val properties =
            Properties().apply {
                name = "Book"
                author = "Auth"
                savePath = tempDir.absolutePath
                addGt = true
                chapList =
                    (1..3).map {
                        Chapter(url = "u$it", chapName = "Chương $it").apply { id = "C$it" }
                    }
            }

        ToC(platformEngineLogger(), properties, autoSplit = false, includeImg = false, settings = settings).mkToC()

        val opf = File(tempDir, "data/content.opf").readText()
        assertTrue(opf.contains("id=\"C0\" href=\"Text/C1.html\""))
        assertTrue(opf.contains("id=\"C2\" href=\"Text/C3.html\""))
        assertTrue(opf.contains("id=\"gioithieu\""))
        // spine: item → itemref
        assertTrue(opf.contains("<itemref idref=\"gioithieu\"/>"))
        // mucluc.html sinh ra
        assertTrue(File(tempDir, "data/Text/mucluc.html").exists())
    }

    @Test
    fun splitPartDetectsChapter1Restart() {
        val chapters =
            listOf(
                Chapter(url = "u1", chapName = "Chương 1").apply { id = "C1" },
                Chapter(url = "u2", chapName = "Chương 2").apply { id = "C2" },
                Chapter(url = "u3", chapName = "Chương 3").apply { id = "C3" },
                Chapter(url = "u4", chapName = "Chương 1").apply { id = "C4" }, // reset → split
                Chapter(url = "u5", chapName = "Chương 2").apply { id = "C5" },
            )
        val properties = Properties().apply { chapList = chapters }

        val splits = ToC(platformEngineLogger(), properties, settings = settings).splitPart()

        // Strategy 3: restart tại index 3
        assertTrue(splits.contains(3), "expected split at chapter-1 restart, got $splits")
    }
}
